### PR TITLE
fix(rpcd): rules map reaches the client; dedupe keys; harden poll clamp

### LINKED ISSUE #217: blocker: rules map lost to pipeline subshell — nft/iptables log-prefix labels never reach the client
## Parent
Part of umbrella #216. Blocks #209.

## Fact
`build_rules_map` in `openwrt-feed/luci-app-fwlive/root/usr/libexec/rpcd/fwlive` assembles the rule-hint
map on the **right-hand side of a pipeline**:

```sh
nft)      nft_list_ruleset | map_from_nft_stream ;;
iptables) iptables-save 2>/dev/null | map_from_iptables_save_stream
          ip6tables-save 2>/dev/null | map_from_iptables_save_stream ;;
```

`map_from_*_stream` → `map_prefix_with_label` → `map_add`, which mutates the shell variable `OUT`.
In POSIX `sh` (BusyBox `ash` / `dash`) the RHS of a pipeline runs in a **subshell**, so every mutation
updates a copy that is discarded on subshell exit. Only `map_uci_rule_names` (run in the current shell)
survives into the emitted JSON.

Net effect: **no nft `log prefix` label, no fw4 `comment "!fw4: …"` label, and no iptables `--log-prefix`
label ever reaches the client** — on both firewall backends.

## Evidence
Executed the real `rules` method with stub `nft`/`uci` on `PATH` (ruleset contained
`log prefix "fwlive-ssh " comment "!fw4: Allow-SSH"`):

```
$ PATH=stub:$PATH dash    .../rpcd/fwlive call rules
{"backend":"nft","rules":{"Allow-DHCP-Renew":"Allow-DHCP-Renew","allow-dhcp-renew":"Allow-DHCP-Renew"}}
$ PATH=stub:$PATH busybox sh .../rpcd/fwlive call rules
{"backend":"nft","rules":{"Allow-DHCP-Renew":"Allow-DHCP-Renew","allow-dhcp-renew":"Allow-DHCP-Renew"}}
```

The `fwlive-ssh` / `Allow-SSH` mappings are absent. Same result for the `iptables` backend.

## Why it shipped — test blind spot
`rulesmap_from_iptables_file` (the `__rulesmap_iptables` CLI selftest) feeds the *same* function through a
**redirect**, which stays in the current shell, so it passes:

```
$ busybox sh .../rpcd/fwlive __rulesmap_iptables
{"backend":"iptables","rules":{"fwlive-wan":"fwlive wan","fwlive-wan":"fwlive wan"}}
```

`tests/fwlive-rules-map.test.js` invokes **only** `__rulesmap_iptables`. There is therefore **zero automated
coverage of the production pipeline path**, which is why an otherwise well-tested package shipped this.

## Impact
`htdocs/luci-static/resources/view/status/fwlive.js` sets `this.rulesMap = (res && res.rules) || {}` and
resolves labels via `rulesMap[hint]` / `rulesMap[slug]`. With the map near-empty, the rule-label feature is
non-functional on a real device for the primary fw4/nft backend. Advertised feature, silent failure.

**Fix location:** `rpcd/fwlive`, `fwlive-logging.sh` and `fwlive-log-filter.sh` are **byte-identical** between `openwrt-feed/luci-app-fwlive/` and the luci PR, so the change lands here and is then re-cut. Do not patch the luci PR tree directly.

## Size
**M** — code change is small (3 call sites + 2 stream functions), but it needs a **new test class** that
exercises `build_rules_map` with stub `nft`/`iptables-save` on `PATH`; no nearest-neighbour test exists.
Est. 90–150 min code+test, plus 60–90 min on-device (QEMU) verification of the nft path. **~2.5–4 h.**

Deps: none (independent of all other children).

## Non-goals
Refactoring the classifier or the log filter (see child C). No behaviour change to `map_uci_rule_names`.

## Acceptance
- [ ] `rules` returns nft `log prefix` and fw4 `comment` labels on a live nft/fw4 device
- [ ] `rules` returns `--log-prefix` labels on a live iptables/fw3 device
- [ ] A test exercises the **production** path (not only the redirect selftest) and fails against current `master`
- [ ] `./scripts/fwlive-test.sh` and `./scripts/validate-baseline.sh` green

---
*Provenance: automated review of [openwrt/luci#8992](https://github.com/openwrt/luci/pull/8992) at commit `91b4629` by the `openwrt-ai` reviewer, independently re-verified locally by **executing** the shipped code under `dash` and BusyBox `ash`. Facts only — no fix or plan in this issue.*

### LINKED ISSUE #221: security: poll lines clamp bypassable via digit-string overflow
## Parent
Part of umbrella #216. Blocks #209.

## Fact
`poll_lines_from_input` in `root/usr/libexec/rpcd/fwlive` validates the caller-supplied line count with a
digit-only check that accepts **any number of digits**, then clamps with a numeric comparison whose failure
is silenced:

```sh
case "$first" in
	''|*[!0-9]*) ;;
	*) lines="$first" ;;
esac

if [ "$lines" -gt "$POLL_LINES_MAX" ] 2>/dev/null; then
	lines="$POLL_LINES_MAX"
fi
```

An over-long digit string overflows `test -gt`; the comparison errors, `2>/dev/null` hides the diagnostic,
the failing test takes the **else** branch, and `lines` reaches the `ubus call log read` JSON **unclamped**.

## Evidence
Both target shells:

```
dash:       input=99999999999999999999   -> forwarded lines=99999999999999999999
dash:       input=18446744073709551616   -> forwarded lines=18446744073709551616
dash:       input=2001                   -> forwarded lines=2000      (clamp works)
dash:       input=0                      -> forwarded lines=0
busybox sh: input=99999999999999999999   -> forwarded lines=99999999999999999999
busybox sh: input=0                      -> forwarded lines=0
```

`POLL_LINES_MAX` is 2000. `{"addresses":["0"]}` forwards `"lines":0`, which is not a value any caller can
meaningfully intend.

## Impact
A **read-ACL** session (the ACL grants `poll` under `read`) can drive
`ubus call log read '{"lines":<huge>}'`, bypassing the intended bound. Practical severity is **medium, not
high**: logd's ring is finite, and an oversized JSON integer may simply error into the existing
`{"log":[]}` fallback. The defect is that a deliberate defence-in-depth clamp does not hold, and the
failure is silent.

**Fix location:** `rpcd/fwlive`, `fwlive-logging.sh` and `fwlive-log-filter.sh` are **byte-identical** between `openwrt-feed/luci-app-fwlive/` and the luci PR, so the change lands here and is then re-cut. Do not patch the luci PR tree directly.

## Size
**S** — reject digit strings longer than `POLL_LINES_MAX` before the numeric comparison, and treat `0` as
"use the default". ~10 lines plus a unit test covering both shells. Est. **30–45 min.**

Deps: none.

## Non-goals
Changing `POLL_LINES_MAX`. Reworking the rpcd argument schema.

## Acceptance
- [ ] Over-long digit strings clamp to `POLL_LINES_MAX` on both `dash` and BusyBox `ash`
- [ ] `0` resolves to the default (50), never forwards `"lines":0`
- [ ] A test asserts the clamp for overflow input and fails against current `master`
- [ ] No reliance on a silenced comparison error for control flow

---
*Provenance: automated review of [openwrt/luci#8992](https://github.com/openwrt/luci/pull/8992) at commit `91b4629` by the `openwrt-ai` reviewer, independently re-verified locally by **executing** the shipped code under `dash` and BusyBox `ash`. Facts only — no fix or plan in this issue.*

### LINKED ISSUE #223: bug: duplicate JSON keys in the rules map when slug_key is an identity
## Parent
Part of umbrella #216.

## Fact
This finding was **not** raised by the upstream reviewer; found during local verification.

`slug_key` is an identity function for already-lowercase, hyphenated input:

```sh
slug_key() {
	echo "$1" | tr '[:upper:]' '[:lower:]' | tr ' _' '--'
}
```

`map_prefix_with_label` unconditionally adds **both** the raw key and its slug:

```sh
map_add "$prefix" "$cosmetic"
map_add "$(slug_key "$prefix")" "$cosmetic"
```

When `slug_key "$prefix" == "$prefix"` (any lowercase, hyphen-only log prefix — the common fw4 case), the
same key is emitted twice.

## Evidence
```
$ busybox sh .../rpcd/fwlive __rulesmap_iptables
{"backend":"iptables","rules":{"fwlive-wan":"fwlive wan","fwlive-wan":"fwlive wan"}}
```

Note the duplicated `"fwlive-wan"` key.

## Impact
Low. Duplicate object keys are tolerated by `JSON.parse` (last value wins), so the UI behaves correctly, but
the emitted JSON is not well-formed in spirit, and the `rules` payload is inflated — up to 2x for
prefix-only maps, on a response sent to every client that opens the view.

**Fix location:** `rpcd/fwlive`, `fwlive-logging.sh` and `fwlive-log-filter.sh` are **byte-identical** between `openwrt-feed/luci-app-fwlive/` and the luci PR, so the change lands here and is then re-cut. Do not patch the luci PR tree directly.

## Size
**S** — skip the second `map_add` when the slug equals the raw key (or make `map_add` idempotent per key).
~5 lines plus a test asserting no duplicate keys in the emitted map. Est. **20–30 min.**

Deps: overlaps the same functions as child A — land **after** A to avoid a conflicting re-cut, or fold the
one-line guard into A's branch.

## Non-goals
Redesigning the key scheme or the slug rules.

## Acceptance
- [ ] `rules` output contains no duplicate JSON keys for lowercase hyphenated prefixes
- [ ] Both raw-key and slug lookups from `fwlive.js` still resolve
- [ ] A test asserts key uniqueness in the emitted map

---
*Provenance: automated review of [openwrt/luci#8992](https://github.com/openwrt/luci/pull/8992) at commit `91b4629` by the `openwrt-ai` reviewer, independently re-verified locally by **executing** the shipped code under `dash` and BusyBox `ash`. Facts only — no fix or plan in this issue.*

### PR DESCRIPTION
**Pull request by lucas-albers-lz4**
fix(rpcd): rules map reaches the client; dedupe keys; harden poll clamp

Three verified defects from the upstream review round on openwrt/luci#8992
(fwlive #217, #221, #223), reproduced locally under `dash` and BusyBox `ash`
before fixing, plus review-gate folds.

Rules map lost to a pipeline subshell (#217)
--------------------------------------------
`build_rules_map` fed `map_from_nft_stream` / `map_from_iptables_save_stream`
from the right-hand side of a pipeline. In POSIX sh that is a subshell, so
every `OUT` mutation was discarded and no nft `log prefix`, fw4 `comment` or
iptables `--log-prefix` label ever reached the client — only the uci-derived
names survived. The existing test only drove `__rulesmap_iptables`, which feeds
the same function through a redirect (current shell) and therefore always
passed.

Dumps are now written with `mktemp` and consumed through a redirect, keeping
accumulation in the main shell. nft prefixes are normalized (fw4 emits
`log prefix "name "` with a trailing space; `parseRuleHint()` yields the bare
token) — normalization is idempotent, so both backends emit the key the client
actually looks up, with no cross-backend drift. Temp files use `mktemp` only
(fixed `/tmp`, `TMPDIR` deliberately not honoured, no predictable fallback),
because this plugin runs as root under rpcd and a guessable path would be a
symlink arbitrary-write primitive. If `mktemp` fails, backend enrichment is
skipped and a well-formed map is still returned.

Duplicate JSON keys in the rules map (#223)
-------------------------------------------
`slug_key` is an identity for already-lowercase hyphenated prefixes, so raw key
and slug were both emitted. `map_add` is now idempotent per key, first-wins:
UCI rule names (added first) beat cosmetic log-prefix labels. The membership
test is anchored on the quoted, json-escaped key, so it is literal (glob
characters match literally) and cannot false-positive on a substring key or on
a key inside a value. Collisions from distinct prefixes whose slugs match are
also suppressed.

Poll line-count clamp was bypassable (#221)
-------------------------------------------
The digit check accepted any length, then `[ "$lines" -gt "$POLL_LINES_MAX" ]`
overflowed on a long digit string; `2>/dev/null` hid the error and the failing
test took the else branch, forwarding an unclamped value to `ubus call log
read` from a read-ACL-reachable method. `{"addresses":["0"]}` also forwarded
`"lines":0`. Clamping now lives in a pure `poll_clamp_lines` helper that
rejects over-long digit strings by length before any numeric comparison and
maps zero to the default; it validates non-digit input and is testable in
`__selftest` without `jshn`.

Tests
-----
The suite now exercises the production `rules` path with stub `nft` /
`iptables-save` on PATH under both `dash` and BusyBox `ash` (the previous test
could not catch the subshell bug), asserts keys exactly as the client looks
them up, asserts no duplicate keys on raw JSON (parsed objects hide them), and
covers backend parity, collisions, glob-character keys, IPv4+IPv6
cross-fragment dedup, mktemp-absent graceful degradation with no predictable
file, and the clamp boundary cases. New tests red-proved against the pre-fix
script. `security-review.md` updated (26 -> 30 controls).

Verified on both shells: nft/iptables labels present with client-matchable
keys, no duplicate or whitespace-padded keys, uci names not regressed,
`__rulesmap_iptables` unchanged, clamp behaviour for 99999999999999999999 /
18446744073709551616 / 2001 / 0 / 500 / 50, TMPDIR ignored, mktemp failure
degrades gracefully. `./scripts/fwlive-test.sh` and
`./scripts/fwlive-shellcheck.sh` green. Reviewed by the luna gate: four rounds
(REQUEST_CHANGES x2, then APPROVE).

Test plan note: stub-verified on both shells; a live-device pass (nft/fw4 and
iptables/fw3) is planned on hardware and tracked in #216.

Closes #217, Closes #221, Closes #223
Refs #216

## Summary by CodeRabbit

* **Bug Fixes**
  * Prevented duplicate firewall rule entries and standardized rule labels.
  * Improved temporary-file security, cleanup, and fallback behavior.
  * Added safeguards for invalid, zero, and oversized polling values.
  * Strengthened temporary-directory validation by requiring sticky permissions and rejecting symlink paths.

* **Tests**
  * Expanded coverage for rule processing, validation, degradation, and temporary-directory security.
  * Improved compatibility across available POSIX shells.

* **Documentation**
  * Updated security documentation with current controls, hardening measures, and privileged-surface restrictions.

### CODE REVIEW

**Comment by lucas-albers-lz4** [comment]
@coderabbitai review

File: docs/developer/security-review.md [hunk 1]
```diff
@@ -303,3 +307,34 @@ lock (#204), removal of unpinned `@playwright/mcp@latest` (#205). Ledger refresh
...
+
+**Fix (minimal, POSIX sh).**
+
+- **Delete fallback:** removed `|| printf '/tmp/fwlive-nft-%s' "$$"` from all 3 sites (no predictable-path code path remains).
+- **Single helper:** `_fwlive_mktemp <prefix>` — `mktemp "/tmp/<prefix>.XXXXXX" 2>/dev/null` only. `TMPDIR` is **deliberately not honoured** and there is **no bare-`mktemp` fallback**: bare `mktemp` consults `TMPDIR`, and an attacker-writable non-sticky `TMPDIR` would restore the create-then-reopen (TOCTOU) symlink-write primitive, since `build_rules_map` reopens the path with `>` while running as root. A fixed sticky `/tmp` removes the vector. Never `rm`+reuse, never `touch`/`chmod`. If `mktemp` fails, the backend enrichment is skipped and a well-formed map is still returned. BusyBox `mktemp` sane.
+- **Graceful degradation:** ` _tmp=$(_fwlive_mktemp fwlive-nft) || _tmp=''` then `if [ -n "$_tmp" ]; then nft_list_ruleset >"$_tmp" ...; rm -f "$_tmp"; fi` — if `mktemp` absent/failing, backend enrichment skipped, still returns well-formed `{"backend":...,"rules":{...}}` with UCI names. No fixed-path write.
```
[the remark this answers is not in the document]
Comment (lucas-albers-lz4) [reply docs/developer/security-review.md]: **FOLD** — leading space in the code span is gone; the assignment is now a double-backtick span so the inner `$()` stays literal.

File: docs/developer/security-review.md [hunk 2]
```diff
@@ -303,3 +307,34 @@ lock (#204), removal of unpinned `@playwright/mcp@latest` (#205). Ledger refresh
...
+- **Cleanup guaranteed:** `rm -f "$_tmp"` is unconditional inside the `if [ -n "$_tmp" ]` block, immediately after use, with no early `return`/`exit` between creation and removal; each prefix gets its own variable (`_tmp`/`_tmp2`) and its own `rm -f`. Failure path (`mktemp` empty) never creates a file, so no cleanup needed. The `map_from_*` helpers do not `exit` the shell.
```
[the remark this answers is not in the document]
Comment (lucas-albers-lz4) [reply docs/developer/security-review.md]: **FOLD** — ledger now matches the code: one `_tmp`, fresh `_fwlive_mktemp` after `rm -f` for nft / iptables / ip6tables. The `_tmp`/`_tmp2` claim was wrong.

**Comment by lucas-albers-lz4** [comment]
Folded the CodeRabbit round on this tip (`250e7e0`):

| Label | Item |
|---|---|
| **FOLD** | MD038 space in the graceful-degradation code span |
| **FOLD** | `_tmp`/`_tmp2` lie — one `_tmp`, reused after `rm -f` |
| **FOLD** | Silent-pass when no POSIX shell — `posixShells()` fails the suite if neither `dash` nor `busybox` is on PATH; mktemp degradation also asserts at least one shell ran |
| **FOLD** | `_fwlive_mktemp` `prefix` global — now `_mk_prefix` |
| **DISMISS** | JS docstring 80% gate — this is POSIX sh + a Node test file |

Task 1 in the ledger now points at R4 (`_fwlive_mktemp` + redirect) instead of the old pipeline-subshell snippet.

**Comment by lucas-albers-lz4** [comment]
@coderabbitai review
